File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
    m := make(map[ListNode]bool)
	for {
        if head == nil {
            return nil
        }
		_, ok := m[*head]
		if !ok {
			m[*head] = true
			head = head.Next
		} else {
			return head
		}
	}
}

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
import java.util.HashSet;

public class Solution {
    public ListNode detectCycle(ListNode head) {
        var set = new HashSet<ListNode>();
        while (head != null) {
            if (!set.contains(head)) {
                set.add(head);
            } else {
                return head;
            }
            head = head.next;
        }
        return null;
    }
}